src/history.rs

```diff
@@ -155,126 +153,184 @@ impl<'a> DynamicHistory<'a> {
-    /// Note: This function silently drop events if the log has
-    /// overflowed or if the instance is locked for reporting.
+    /// Note: this function overwrites older events in the log if it
+    /// is full.
     #[inline]
     pub(crate) fn record_event_with_payload(&mut self, event_id: EventId, payload: u32) {
-        if self.chunked_report_state.is_report_in_progress() {
-            return;
-        }
-        let len = self.compact_log.len();
-        let cap = self.compact_log.capacity();
-        // Room for two?
-        if len + 1 >= cap {
-            return;
-        }
-        let (ev, payload) = CompactLogItem::event_with_payload(event_id, payload);
-        if self.compact_log.try_push(ev).is_err() {
-            return;
-        }
-        if self.compact_log.try_push(payload).is_ok() {
-            self.event_count = self.event_count.saturating_add(1);
-        }
+        let (ev, pay) = LogEntry::event_with_payload(event_id, payload);
+        self.write_double_to_log(ev, pay);
+        self.event_count = self.event_count.saturating_add(1);
     }
 
     /// Increments the clock in the logical clock corresponding to this probe instance
     #[inline]
-    fn increment_local_clock_count(&mut self) {
+    fn increment_local_clock(&mut self) {
         // N.B. We rely on the fact that the first member of the clocks
         // collection is always the clock for this probe
-        debug_assert!(self.probe_id == self.clocks[0].id);
-        debug_assert!(
-            !self.chunked_report_state.is_report_in_progress(),
-            "Should not be incrementing the local clock count when a report is in progress"
-        );
-        self.clocks.as_mut_slice()[0].increment();
+        self.self_clock.increment();
         self.event_count = 0;
     }
 
     #[inline]
-    pub(crate) fn produce_snapshot(&mut self) -> Result<CausalSnapshot, ProduceError> {
-        if self.chunked_report_state.is_report_in_progress() {
-            return Err(ProduceError::ReportLockConflict);
+    pub(crate) fn now(&self) -> ModalityProbeInstant {
+        ModalityProbeInstant {
+            clock: self.self_clock,
+            event_count: self.event_count,
+        }
+    }
+
+    pub(crate) fn report(&mut self, destination: &mut [u8]) -> Result<usize, ReportError> {
+        // If we can't store at least a header and one event, it's a hard error
+        if destination.len() < WireReport::<&[u8]>::buffer_len(0, 1) {
+            return Err(ReportError::InsufficientDestinationSize);
+        }
+
+        let self_clock = self.self_clock;
+        let mut report = WireReport::new_unchecked(destination);
+        report.set_fingerprint();
+        report.set_probe_id(self.probe_id);
+        report.set_clock(crate::pack_clock_word(self_clock.epoch, self_clock.ticks));
+
+        // We can't store at least the frontier clocks and a single two-word item
+        if report.as_ref().len() < WireReport::<&[u8]>::buffer_len(self.clocks.len(), 2) {
```

Comment:
In this first branch, I think we need only one word-sized item, the event id of the "not enough space" event.

Reply:
A very similar check at the top returns an error. I think we should probably be consistent and do the same.

I understand it's a trade-off between being able to produce a report that contains an error-signaling-event (which might make it to the collector for analysis) versus producing a local error, and in this case I think the local error is preferable because the remediation is certainly going to be locally-specific, and also not particularly difficult.